### PR TITLE
Update package and module names for Modules V2

### PR DESCRIPTION
**Pull request by FloLey**
- make sure - and _ are converted correctly in the module name: the Python package name should use - while all other names (even the Python module name and directory) should use _.

- make sure capital letters are dealt with:

   - a capital letter at the start of a module name should be converted to a lower case letter (give a warning to the user)

   - capital letters in other positions of the module name should not be converted

issue #14

### CODE REVIEW

**Comment by FloLey** [comment]
@sanderr 
I tried to combine what we said and what was in the ticket but I'm not sure everything is correct. Especially concerning the name of the root folder which is now exactly what the user inputs.( I had to change the jenkinsfile so that the test would pass).
I also have some trouble with setting the warning message as the "checks and changes" are done in the pre-hook.
I tried to print some things there but (in my opinion) those prints were lost with due to the outputs of the post-hook. I was not able to find how to retrieve the context in the output .sh file which would have help me for a more visible message. therefore I now hid the outputs of the git commands and I print one custom message. What do you think about that?

**Comment by sanderr** [comment]
I'm going to postpone the detailed review to tomorrow. For now I'll say it's not too bad in my opinion if we don't show a warning. The key part is that the conversion happens so that the end result is a valid module. I'm not sure about hiding the git output but I see what you mean about it obfuscating the intended output. I guess as long as we only redirect stdout and not stderr it might be acceptable. I'll have to double check git's messages about email address etc being unset are actually output to stderr though, we don't want to hide those.

File: hooks/pre_gen_project.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{{ cookiecutter.update({"package_name": "inmanta-module-"+cookiecutter.module_name.lower().replace("_","-") }) }}
```
Comment (sanderr) [inline hooks/pre_gen_project.py:1]: Could you add a description at the top of the file explaining the purpose of this script and the semantics (what they represent as well as what naming conventions they follow) of both variables it introduces?

File: hooks/pre_gen_project.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
...
+{{ cookiecutter.update({"inmanta_module_name": (cookiecutter.module_name[0].lower()+cookiecutter.module_name[1:]).replace("-","_") }) }}
+
+print("===========================")
+print("Module name: {{ cookiecutter.inmanta_module_name }}")
+print("Python package name: {{ cookiecutter.package_name }}")
```
Comment (sanderr) [inline hooks/pre_gen_project.py:6]: After thinking about this some more I think I'm in favor of just converting the name silently, without this message, and to leave the git output as is. What do you think?

File: Jenkinsfile [hunk 3]
```diff
@@ -21,7 +21,7 @@ pipeline {
                         ${WORKSPACE}/env/bin/cookiecutter --no-input ../inmanta-module-template/
                     '''
                 }
-                dir('module/inmanta-module-test-module') {
+                dir('module/test_module') {
```
Comment (sanderr) [inline Jenkinsfile:24]: I'm going back and forth on whether we should use for the top level directory name the module name exactly as the user input it or enforce a convention. At the moment I'm back in favor of enforcing the convention (if the user inputs `Test_Module`, convert it to `test-module`), just to enforce consistency (because 99% this directory name will be the repo name as well).
@wouterdb @arnaudsjs do you have any opinion on this?

**Review by sanderr — CHANGES_REQUESTED** [review]
I can't attach comments to file renames so I'll note it here:
- the plugins directory (`inmanta_plugins/<module-name>`) needs to use the Python module name, which is the same as the inmanta module name (`inmanta_module_name`). Related, this will require updating `MANIFEST.in`.
- The README contains the module name as well, I think this should be the inmanta module name.

Apart from this, and the two things you already started a discussion on, all looks good to me. Don't forget to open a pull request on the v1 module template once this one gets approved.

File: hooks/pre_gen_project.py (hunk 2 above)
Comment (FloLey) [reply hooks/pre_gen_project.py:6]: I'm not particularly attached to those outputs :p I'll remove them.

File: hooks/pre_gen_project.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+    'Inmanta module name' using the module name inputed by the user (or the default module name)
+    The Python package name uses the same naming convention as standart Python packages: 
+    all lowercase and words separated by "_".
+    The 'Inmanta module name' (which is also the name of the Python module name) uses "-" to separate
+    words and can have uppercase letters except for the first letter which is always lowercase.
+
+"""
+
+{{ cookiecutter.update({"package_name": "inmanta-module-" + cookiecutter.module_name.lower().replace("_","-") }) }}
+{{ cookiecutter.update({"inmanta_module_name": (cookiecutter.module_name[0].lower() + cookiecutter.module_name[1:]).replace("-","_") }) }}
```
Comment (sanderr) [inline hooks/pre_gen_project.py:12]: This will fail to convert the first character if it happens to be a `-`. But if the user inputs something that exotic they deserve what's coming imo.

File: hooks/pre_gen_project.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+"""
+    This script has as purpose to create the Python package name and the 
+    'Inmanta module name' using the module name inputed by the user (or the default module name)
+    The Python package name uses the same naming convention as standart Python packages: 
+    all lowercase and words separated by "_".
+    The 'Inmanta module name' (which is also the name of the Python module name) uses "-" to separate
```
Comment (sanderr) [inline hooks/pre_gen_project.py:6]: You've got the `-` and `_` turned around here.

File: hooks/pre_gen_project.py (hunk 4 above)
Comment (FloLey) [reply hooks/pre_gen_project.py:12]: maybe like this?

```
{{ cookiecutter.update({"inmanta_module_name": (cookiecutter.lstrip(string.punctuation).module_name[0].lower() + cookiecutter.lstrip(string.punctuation).module_name[1:]).replace("-","_") }) }}
```

I did not find how to use variables declared out off the scope of the {{ }} in the {{ }}, which would help make it more readable :/

Comment (sanderr) [reply hooks/pre_gen_project.py:12]: Actually, I wouldn't pay too much attention to this, I'm pretty sure a module named `_my_module` would be rejected by the compiler as well so I think for now we can just keep assuming the user will input something at least somewhat sane. You can leave this as you initially wrote it in my opinion.

File: Jenkinsfile (hunk 3 above)
Comment (arnaudsjs) [reply Jenkinsfile:24]: I am also in favor of doing the conversion.

File: {{cookiecutter.module_name|slugify}}/.gitignore [hunk 6]
```diff
@@ -1,5 +1,5 @@
-inmanta_plugins/**/__pycache__
-tests/**/__pycache__
+plugins/__pycache__
+tests/__pycache__
```
Comment (arnaudsjs) [inline {{cookiecutter.module_name|slugify}}/.gitignore:2]: Why is this change required?

File: hooks/pre_gen_project.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+"""
+    This script has as purpose to create the Python package name and the 
+    'Inmanta module name' using the module name inputed by the user (or the default module name).
```
Comment (arnaudsjs) [inline hooks/pre_gen_project.py:3]:
```suggestion
    'Inmanta module name' using the module name inputted by the user (or the default module name).
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: hooks/pre_gen_project.py (hunk 7 above)
Comment (sanderr) [reply hooks/pre_gen_project.py:3]: I believe plain `input` is what you're looking for.

File: {{cookiecutter.module_name|slugify}}/.gitignore (hunk 6 above)
Comment (sanderr) [reply {{cookiecutter.module_name|slugify}}/.gitignore:2]: This looks like a mistake in merging v2 into this branch.

**Review by sanderr — APPROVED** [review]
Something seems to have gone wrong when merging v2 into this branch but @arnaudsjs already remarked on that change. The rest looks good to me.

**Review by arnaudsjs — APPROVED** [review]
